Add HomePage and display restaurants

rachelekm opened the pull request:
## Overview

This creates a public homepage at route "/" that displays restaurants within the bounds of the Map bounding box. Restaurants are plotted on the Map and their information is displayed in the Sidebar. On Map move, restaurants and their info should synchronously update according to the new bounding box and similarly update when a user interacts with pagination in the Sidebar: "next", "previous", or selecting page number. 
Connects #4 

### Demo

HomePage with paginated restaurants displayed in Map and Sidebar:
![Screen Shot 2021-09-07 at 1 58 26 PM](https://user-images.githubusercontent.com/36374797/132391744-0a138642-2f68-4fd2-9531-c05cbe149cf7.png)
On selecting "next" or page 2, Map and Sidebar update restaurants:
![Screen Shot 2021-09-07 at 1 58 41 PM](https://user-images.githubusercontent.com/36374797/132391835-effeb2f7-f303-4f49-9895-00086e302298.png)


### Notes

I deleted frontend auth functionality (including: fetch to get the current User info, PrivateRoute, route to "/auth") since it doesn't fit current needs after making HomePage public and had the potential to create ts unused variable errors. Will eventually have to add this back in for authorizing users on sign in and to leave ratings.

## Testing Instructions

 * Start server and open http://localhost:4567 in browser
 * Confirm HomePage displays on route "/"
 * Confirm that restaurants are paginated and display on both Map and Sidebar
 * Confirm that paginated restaurants update on Map move as well as on "next", "previous", and when selecting page number in Sidebar

rachelekm committed:
Add map and restaurants to HomePage

- Updated store and added actions/reducers/api request for fetching restaurants according to map bounds
- Django updates to respond to api request for restaurants with geojson object

rachelekm committed:
Created Map component

rachelekm committed:
Added Sidebar component

rachelekm committed:
Add average_rating field and test

rachelekm committed:
Paginate Restaurants

- frontend paginate Restaurant data and add mapData to store
- backend add rest_framework pagination and ordering by name

rachelekm committed:
Create tests for RestaurantSerializer and RestaurantList

rachelekm committed:
Make HomePage a public route

rachelekm committed:
Remove unnecessary frontend auth

maurizi commented on client/src/actions/restaurants.ts:
```diff
@@ -0,0 +1,7 @@
+import { PaginatedRestaurants } from "../models";
+import { LngLatBounds } from "mapbox-gl";
+
+export const restaurantsFetch = createAction("Restaurants Fetch")<[LngLatBounds | null, number]>();
```
Should this be the following?
```suggestion
export const restaurantsFetch = createAction("Restaurants Fetch")<[LngLatBounds, number]>();
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+        map.touchZoomRotate.disableRotation();
+        map.doubleClickZoom.disable();
+
+        return () => map?.remove();
```
Cleaning up like this is a good habit to pick up 🎉

maurizi commented on django/api/factories.py:
```diff
@@ -0,0 +1,25 @@
+    name = factory.Sequence(lambda n: "Test Restaurant {}".format(n))
+    address = "123 Street St. Place, PA 12345"
+    loc = Point(0, 0)
+    is_approved = True
```
Nice job setting up these factories - that will make adding more tests a lot easier 🎉 

FYI it turns out the 12345 zip code is in NY.

maurizi commented on client/src/reducers/mapdata.ts:
```diff
@@ -0,0 +1,36 @@
+
+import { mapData } from "../models";
+
+export interface mapDataState {
```
Conventionally, types should be spelled in CapitalCase, while variables should be spelled with camelCase, so this should be `mapDataState`.

maurizi commented on client/src/models.ts:
```diff
@@ -36,3 +38,31 @@ export interface User {
+    readonly average_rating: number | null;
+}
+
+type NoBbox<T> = Omit<T, "bbox" | "id">;
```
Given the name of `NoBbox` I'd expect this to _just_ be an alias for `Omit<T, "bbox">`.

Note: this change on it's own will cause a type error bc `featureDecoder` in `decoders.ts` doesn't know about `id` - you will need to update the decoder as well.
```suggestion
type NoBbox<T> = Omit<T, "bbox">;
```

maurizi commented on client/src/components/Sidebar.tsx:
```diff
@@ -0,0 +1,85 @@
+
+    useEffect(() => {
+        mapData.data.bounds && currentPag && store.dispatch(restaurantsFetch([mapData.data.bounds, currentPag]));
+        // eslint-disable-next-line
```
It's in general better to be specific about the lint warning you're suppressing:
```suggestion
        // eslint-disable-next-line react-hooks/exhaustive-deps
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+
+        return () => map?.remove();
+
+        // eslint-disable-next-line
```
```suggestion
        // eslint-disable-next-line react-hooks/exhaustive-deps
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+    }, [initializedMap, allRestaurants]);
+
+    useEffect(() => {
+        // eslint-disable-next-line
```
No lint here?
```suggestion
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+        if (mapRef.current === null) {
+            return;
+        }
+        // eslint-disable-next-line
```
This doesn't seem to be causing a lint warning for me?
```suggestion
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+            });
+
+            const handlePopupEvent = (e: mapboxgl.MapLayerMouseEvent) => {
+                // eslint-disable-next-line
```
This doesn't seem to be causing a lint warning for me?
```suggestion
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+            };
+
+            initializedMap.on("mouseenter", "restaurants-layer", function (e: mapboxgl.MapLayerMouseEvent) {
+                // eslint-disable-next-line
```
This doesn't seem to be causing a lint warning for me?
```suggestion
```

maurizi commented on client/src/components/Map.tsx:
```diff
@@ -0,0 +1,174 @@
+            });
+
+            initializedMap.on("mouseleave", "restaurants-layer", function () {
+                // eslint-disable-next-line
```
This doesn't seem to be causing a lint warning for me?
```suggestion
```

maurizi reviewed (commented):
Mostly looking really good & polished 🎉 

I left some smaller comments but overall this is great.

One thing that might be nice to pair on is the styling - you can see the sidebar is a bit off, and  in general using `position: absolute` & `vw` / `vh` units is usually not the best way for these kinds of whole page elements.
![image](https://user-images.githubusercontent.com/4432106/133358041-3efbad87-7102-4f8a-b3b6-6e55871eebf6.png)

rachelekm committed:
Revise mapData types

In order to ensure mapData.bounds parameter is not type null for fetchRestaurants, set default bounds in mapData and allow mapData.zoom and mapData.center null types to be reset on map load

rachelekm committed:
Modify NoBbox type to include id

rachelekm committed:
Lint, format, and styling fixes

rachelekm committed:
Fixing styling and formatting

- Frontend css flexbox
- Fixing CaptialCase errors for types
- Adding CHOKIDAR_USEPOLLING in start script

rachelekm committed:
Update restaurantsFetch types

maurizi reviewed (approved):
Looks good 👍
